arch/src/aarch64/fdt.rs

```diff
@@ -0,0 +1,424 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use byteorder::{BigEndian, ByteOrder};
+use libc::{c_char, c_int, c_void};
+use std::ffi::{CStr, CString, NulError};
+use std::ptr::null;
+use std::{io, result};
+
+use super::get_fdt_addr;
+use super::gic::{KVM_VGIC_V3_DIST_SIZE, KVM_VGIC_V3_REDIST_SIZE};
+use super::layout::{FDT_MAX_SIZE, MAPPED_IO_START};
+use aarch64::fdt::Error::CstringFDTTransform;
+use memory_model::{GuestAddress, GuestMemory, GuestMemoryError};
+
+// This is a value for uniquely identifying the FDT node declaring the interrupt controller.
+const GIC_PHANDLE: u32 = 1;
+// Read the documentation specified when appending the root node to the FDT.
+const ADDRESS_CELLS: u32 = 0x2;
+const SIZE_CELLS: u32 = 0x2;
+// Taken from qemu.
+const ARCH_GIC_MAINT_IRQ: u32 = 9;
+
+// As per kvm tool and
+// https://www.kernel.org/doc/Documentation/devicetree/bindings/interrupt-controller/arm%2Cgic.txt
+// Look for "The 1st cell..."
+const GIC_FDT_IRQ_TYPE_PPI: u32 = 1;
+
+// From https://elixir.bootlin.com/linux/v4.9.62/source/include/dt-bindings/interrupt-controller/irq.h#L17
+const IRQ_TYPE_LEVEL_HI: u32 = 4;
+
+// This links to libfdt which handles the creation of the binary blob
+// flattened device tree (fdt) that is passed to the kernel and indicates
+// the hardware configuration of the machine.
+extern "C" {
+    fn fdt_create(buf: *mut c_void, bufsize: c_int) -> c_int;
+    fn fdt_finish_reservemap(fdt: *mut c_void) -> c_int;
+    fn fdt_begin_node(fdt: *mut c_void, name: *const c_char) -> c_int;
+    fn fdt_property(fdt: *mut c_void, name: *const c_char, val: *const c_void, len: c_int)
+        -> c_int;
+    fn fdt_end_node(fdt: *mut c_void) -> c_int;
+    fn fdt_open_into(fdt: *const c_void, buf: *mut c_void, bufsize: c_int) -> c_int;
+    fn fdt_finish(fdt: *const c_void) -> c_int;
+    fn fdt_pack(fdt: *mut c_void) -> c_int;
+}
+
+#[derive(Debug)]
+pub enum Error {
+    AppendFDTNode(io::Error),
+    AppendFDTProperty(io::Error),
+    CreateFDT(io::Error),
+    CstringFDTTransform(NulError),
+    FinishFDTReserveMap(io::Error),
+    IncompleteFDTMemoryWrite,
+    WriteFDTToMemory(GuestMemoryError),
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+// Creates the flattened device tree for this VM.
+pub fn create_fdt(guest_mem: &GuestMemory, num_cpus: u32, cmdline: &CStr) -> Result<(Vec<u8>)> {
+    // Alocate stuff necessary for the holding the blob.
+    let mut fdt = vec![0; FDT_MAX_SIZE];
+    // Safe since we allocated this array with FDT_MAX_SIZE.
+    let mut fdt_ret = unsafe { fdt_create(fdt.as_mut_ptr() as *mut c_void, FDT_MAX_SIZE as c_int) };
+
+    if fdt_ret != 0 {
+        return Err(Error::CreateFDT(io::Error::last_os_error()));
+    }
+
+    // The flattened device trees created with fdt_create() contains a list of
```

Comment:
The first line of comments starts with a space, while the other don't. Is this intended ?

Reply:
Right, solved.